crates/biome_js_analyze/src/lint/nursery/no_common_js.rs

```diff
@@ -0,0 +1,153 @@
+use biome_analyze::{context::RuleContext, declare_lint_rule, Rule, RuleDiagnostic, RuleSource};
+use biome_console::markup;
+use biome_js_syntax::{
+    global_identifier, JsCallExpression, JsReferenceIdentifier, JsStaticMemberAssignment,
+};
+use biome_rowan::{declare_node_union, AstNode};
+
+use crate::services::semantic::Semantic;
+
+declare_lint_rule! {
+    /// Disallow use of CommonJs module system in favor of ESM style imports.
+    ///
+    /// ESM-style `import`s are modern alternative to CommonJS `require` imports. Supported by all modern browsers and Node.js versions.
+    /// Tooling can more easily statically analyze and tree-shake ESM `import`s compared to CommonJs.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// require('node:fs');
+    /// ```
+    /// ```js,expect_diagnostic
+    /// module.exports = { a: 'b' }
+    /// ```
+    /// ```js,expect_diagnostic
+    /// exports.a = 'b';
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// import fs from 'node:fs';
+    /// ```
+    /// ```js
+    /// import('node:fs')
+    /// ```
+    /// ```js
+    /// export const a = 'b';
+    /// ```
+    /// ```js
+    /// export default { a: 'b' };
+    /// ```
+    ///
+    /// ## Caveats
+    ///
+    /// Rule is automatically disabled inside `.cjs` and `.cts` files, because they are explicitly CommonJs files.
+    ///
+    /// This rule could be helpful if you are migrating from CommonJs to ESM,
+    /// but if you wish to continue using CommonJs, you can safely disable it.
+    ///
+    pub NoCommonJs {
+        version: "1.9.0",
+        name: "noCommonJs",
+        language: "js",
+        sources: &[
+            RuleSource::EslintTypeScript("no-require-imports"),
+            RuleSource::EslintImport("no-commonjs"),
+        ],
+        recommended: false,
+    }
+}
+
+impl Rule for NoCommonJs {
+    type Query = Semantic<CommonJsImportExport>;
+    type State = bool;
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let file_ext = ctx.file_path().extension().and_then(|ext| ext.to_str())?;
+        // cjs and cts files can only use CommonJs modules
+        if file_ext == "cjs" || file_ext == "cts" {
+            return None;
+        }
```

Comment:
We could avoid checking whether the extension is in Unicode by just comparing bytes.

I could also avoid returning prematurely if we don't manage to retrieve the extension.
Note that this should never happen. However, this looks cleaner to me.

```suggestion
        let file_ext = ctx.file_path().extension();
        // cjs and cts files can only use CommonJs modules
        if file_ext.is_some_and(|file_ext| matches!(file_ext.as_encoded_bytes(), b"cjs" | b"cts")) {
            return None;
        }
```